tensorflow/core/platform/default/logging.cc

```diff
@@ -31,28 +31,41 @@ limitations under the License.
 #include <string.h>
 #include <time.h>
 
-#include <string>
+#include <algorithm>
+#include <mutex>
+#include <queue>
 #include <unordered_map>
+#include <vector>
 
 namespace tensorflow {
+namespace internal {
 
-void TFAddLogSink(TFLogSink* sink) {
-  // LogSink is not implemented.
-  // If necessary, one can add the log sink support as follows.
-  // 1. Define a global vector<TFLogSink> to keep track of all registered
-  //    TFLogSink objects. Protect the global vector with mutex to make it
-  //    thread-safe.
-  // 2. Add/remove elements from the global vector<TFLogSink> in TFAddLogSink
-  //    and TFRemoveLogSink function
-  // 3. Add logic in LogMessage::GenerateLogMessage() below to dispatch log
-  //    messages to all the registered log sinks.
-}
+class TFLogSinks {
+  public:
+    static TFLogSinks& Instance();
 
-void TFRemoveLogSink(TFLogSink* sink) {
```

Comment:
Please keep the existing API the same (TF{Add,Remove}LogSink). It can forward to your proposed implementation.